File: 0015-3sum/0015-3sum.py
class Solution(object):
    def threeSum(self,nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        i=0
        result=[]
        while i <len(nums)-2:
            if i>0 and nums[i]==nums[i-1]:
                i+=1
                continue
            find=-nums[i]
            j=i+1
            k=len(nums)-1
            while j<k:
                sum=nums[j]+nums[k]
                if sum>find:
                    k-=1
                    
                elif sum<find:
                    j+=1
                    
                else:
                    result.append([nums[i],nums[j],nums[k]])
                    j+=1
                    k-=1
                    while nums[k]==nums[k+1] and k>=0:
                        k-=1
                    while nums[j]==nums[j-1] and j<len(nums)-1:
                        j+=1
            i+=1
        return result

        
                 
                 

File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> numbers=new ArrayList<>();
        Arrays.sort(nums);
        for(int i=0;i<nums.length;i++){
            int target=nums[i];
            int left=i+1;
            int right=nums.length-1;
            if(i>0 && nums[i]==nums[i-1]){
                continue;
            }
            
            while(left<right){
                List<Integer> ans=new ArrayList<>();
                if(nums[left]+nums[right]+target==0){
                    ans.add(target);
                    ans.add(nums[left]);
                    ans.add(nums[right]);
                    numbers.add(ans);
                    left++;
                    right--;
                    while(left<nums.length&&nums[left]==nums[left-1]) {
                        left++;
                    }
                    
                    while(right>=0 && nums[right]==nums[right+1] ){
                        right--;
                    }
                }else if(nums[left]+nums[right]+target<0){
                    left++; 
                }else if(nums[left]+nums[right]+target>0){
                    right--;
                }
            }
        }

        return numbers;
    }
}